js/pro/base/Cache.js

```diff
@@ -207,8 +209,118 @@ class ArrayCacheBySymbolById extends ArrayCache {
     }
 }
 
+class ArrayCacheBySymbolBySide extends ArrayCache {
+
+    constructor () {
+        super ()
+        this.nestedNewUpdatesBySymbol = true
+        Object.defineProperty (this, 'hashmap', {
+            __proto__: null, // make it invisible
+            value: {},
+            writable: true,
+        })
+    }
+
+    append (item) {
+        const bySide = this.hashmap[item.symbol] = this.hashmap[item.symbol] || {}
+        if (item.side in bySide) {
+            const reference = bySide[item.side]
+            if (reference !== item) {
+                for (const prop in item) {
+                    reference[prop] = item[prop]
+                }
+            }
+            item = reference
+            const index = this.findIndex ((x) => x.symbol === item.symbol && x.side === item.side)
+            // move the order to the end of the array
+            this.splice (index, 1)
+        } else {
+            bySide[item.side] = item
+        }
+        this.push (item)
+        if (this.clearAllUpdates) {
+            this.clearAllUpdates = false
+            this.clearUpdatesBySymbol = {}
+            this.allNewUpdates = 0
+            this.newUpdatesBySymbol = {}
+        }
+        if (this.newUpdatesBySymbol[item.symbol] === undefined) {
+            this.newUpdatesBySymbol[item.symbol] = new Set ()
+        }
+        if (this.clearUpdatesBySymbol[item.symbol]) {
+            this.clearUpdatesBySymbol[item.symbol] = false
+            this.newUpdatesBySymbol[item.symbol].clear ()
+        }
+        // in case an exchange updates the same order id twice
+        const sideSet = this.newUpdatesBySymbol[item.symbol]
+        const beforeLength = sideSet.size
+        sideSet.add (item.side)
+        const afterLength = sideSet.size
+        this.allNewUpdates = (this.allNewUpdates || 0) + (afterLength - beforeLength)
+    }
+}
+
+
+class PositionsCache {
```

Comment:
@pcriadoperez this implementation is not very "similar" to the other caches and provides a different API which might be a bit confusing. Moreover, by not extending from one of the Cache classes there are a lot of properties (getLimit, newUpdates, etc) that won't be available.

I don't have all the context here, (so sorry in advance if this is no sense) but couldn't we mimic `watchOrders` approach? I mean the difference is that instead of an `id` we would have a `side` (we can't have multiple positions on the same side), but other than that the update flow seems similar to me.

What do you think?

Reply:
Hey @carlosmiei, 
So that was my initial thought and that's why I implemented `ArrayCacheBySymbolBySide` that you can see in the PR.
However I found a big difference between `watchOrders` and `watchPostions`. `watchOrders` takes in one symbol, while `watchPostions` can take in an array of symbols.
And given that one of the requirements we have is that we want each `watch` function to act independently. So for example if we have newUpdates=true and a user decides to call `watchPositions  (['BTC/USD'])` and then `watchPositions(['BTC/USD', 'ETH/USD'])`. We would have two return the latest updates for BTC/USD in both function calls. This is something that doesn't happen with `watchOrders`

That is why I created `PositionsCache` Where it uses the symbols requested in `watchPositions` as a signature and tracks what functions have seen what updates to allow each `watchPositions` function to act independently.

Also I think a similar approach to this can be used for whenever we implement `watchTickers` which has the same characteristics.

Let me know what you think

Reply:
@pcriadoperez I'm not sure that `PositionCache` fully fixes the issue. 

Let's assume you have this:

```Javascript
const promiseA = exchange.watchPositions(["BTC/USDT"])
const promiseB = exchange.watchPositions(["BTC/USDT","LTC/USDT"])
Promise.all([promiseA,promiseB])
```

If I understood correctly, upon getting an update for the "LTC/USDT" position we will resolve promiseB with the expected result but promiseA will also be resolved with an empty array, which is not expected. 

@pcriadoperez Am I thinking right?

If so, what if we started by implementing `watchPosition` similarly to `watchOrders` and separately try to find a generic solution that would fix this and `watchTickers` "problem"?

Reply:
@carlosmiei Excellent excellent point!

So I think they are two things, one is keeping track of the newUpdates per function, which I think this `PositionCache` does work and the other is knowing when to resolve the promise.

For resolving the promises I can see two different solutions.
#### Option 1 - In the `watchPositions `add:
``` javascript 
while (positions.length === 0) {
    positionsCache = await this.watch(...)
    positions = positionsCache .toArray (symbols, this.newUpdates);
 }
 ```
#### Option 2 - checking messageHash in handlePositions
   - in `watchPositions` store all symbols in messageHash
   - in `handlePositions` loop through client.futures and resolve futures that start with 'positions:' and contain symbol in messageHash

#### Option 3:    
As you said a third option is to not to do `watchPositions` and only do `watchPosition`.

#### Option 4:
And just to brainstorm a forth option is to have `watchPositions()` without a symbols and the user can do the filtering or use watchPosition() for a unique symbol.


In my opinion, not giving `watchPositions()`  to the user when we can seems bad. So I would go with either Option 1 or 2. Or do option 4, which could allow to use a solution like `ArrayBySymbolBySide`. And in that case I would suggest to do the same with `watchTickers()` and we can always let the user do the filtering.

Let me know what you think

Reply:
sorry guys if my comment doesn't make sense (i haven't worked in WS internals extensively, so I am not much in context), but when user subscribes to multiple i.e. `await ex.watchTickers( ['BTC/USDT', 'ETH/USDT'])` then whenever any of them has update i.e. eth, then the response should be similar to fetchTickers, but containing only updated object, like:
```
{
    'ETH/USDT': {
        ....
    }
}
```
as any other behavior would be incorrect in my opinion.

Reply:
@ttodua correct, that's the desired behavior.